services/bonita-business-data/bonita-business-data-impl/src/main/java/org/bonitasoft/engine/business/data/impl/JPABusinessDataRepositoryImpl.java

```diff
@@ -56,35 +60,78 @@
-            final Map<String, Object> configuration) {
+    public JPABusinessDataRepositoryImpl(final TransactionService transactionService, final BusinessDataModelRepository businessDataModelRepository, TechnicalLoggerService loggerService,
+                                         final Map<String, Object> configuration, ClassLoaderService classLoaderService, long tenantId) {
         this.transactionService = transactionService;
         this.businessDataModelRepository = businessDataModelRepository;
-        this.configuration = new HashMap<String, Object>(configuration);
+        this.loggerService = loggerService;
+        this.classLoaderService = classLoaderService;
+        this.tenantId = tenantId;
+        this.configuration = new HashMap<>(configuration);
         this.configuration.put("hibernate.ejb.resource_scanner", InactiveScanner.class.getName());
     }
 
     @Override
     public void start() {
         if (businessDataModelRepository.isDBMDeployed()) {
-            entityManagerFactory = Persistence.createEntityManagerFactory(BDR_PERSISTENCE_UNIT, configuration);
+            loggerService.log(getClass(), TechnicalLogSeverity.DEBUG, "Create entity factory on tenant " + tenantId);
+            entityManagerFactory = createEntityManagerFactory();
         }
+        classLoaderService.addListener(ScopeType.TENANT.name(), tenantId, this);
+    }
+
+    EntityManagerFactory createEntityManagerFactory() {
+        return Persistence.createEntityManagerFactory(BDR_PERSISTENCE_UNIT, configuration);
     }
 
     @Override
     public void stop() {
-        if (entityManagerFactory != null) {
-            entityManagerFactory.close();
+        if (getEntityManagerFactory() != null) {
+            loggerService.log(getClass(), TechnicalLogSeverity.DEBUG, "Close entity factory because service is stopping on tenant " + tenantId);
+            getEntityManagerFactory().close();
             entityManagerFactory = null;
         }
+        classLoaderService.removeListener(ScopeType.TENANT.name(), tenantId, this);
+    }
+
+    private synchronized void recreateEntityManagerFactory(ClassLoader newClassLoader) {
+        if (businessDataModelRepository.isDBMDeployed()) {
+            loggerService.log(getClass(), TechnicalLogSeverity.DEBUG, "Recreate entity factory for classloader " + newClassLoader + " on tenant " + tenantId);
+            ClassLoader currentClassLoader = Thread.currentThread().getContextClassLoader();
+            try {
+                Thread.currentThread().setContextClassLoader(newClassLoader);
+                entityManagerFactory.close();
+                entityManagerFactory = createEntityManagerFactory();
+            } finally {
+                Thread.currentThread().setContextClassLoader(currentClassLoader);
+            }
+        }
+    }
+
+    public EntityManagerFactory getEntityManagerFactory() {
+        if (entityManagerFactory == null) {
```

Comment:
agreed, will first change the get to see if it's closed or nor, but we might have issues when the entity manager is closed after we get them and we execute query on it.
Not sure on what to do here.